//
// Copyright 2021 The Feast Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//


syntax = "proto3";

package feast.core;
option java_package = "feast.proto.core";
option java_outer_classname = "SavedDatasetProto";
option go_package = "github.com/feast-dev/feast/sdk/go/protos/feast/core";

import "google/protobuf/timestamp.proto";
import "feast/core/FeatureViewProjection.proto";
import "feast/core/DataSource.proto";

message SavedDatasetSpec {
  // Name of the dataset. Must be unique since it's possible to overwrite dataset by name
  string name = 1;

  // Name of Feast project that this Dataset belongs to.
  string project = 2;

  // Represents a projection that's to be applied on top of the FeatureView.
  // Contains data such as the features to use from a FeatureView.
  repeated FeatureViewProjection features = 3;

  SavedDatasetStorage storage = 4;
}

message SavedDatasetStorage {
  oneof kind {
    DataSource.FileOptions file_storage = 4;
    DataSource.BigQueryOptions bigquery_storage = 5;
    DataSource.RedshiftOptions redshift_storage = 6;
  }
}

message SavedDatasetMeta {
  // Time where this Saved Dataset is created
  google.protobuf.Timestamp created_timestamp = 1;

  // Time where this Saved Dataset is last updated
  google.protobuf.Timestamp last_updated_timestamp = 2;
}

message SavedDataset {
  SavedDatasetSpec spec = 1;
  SavedDatasetMeta meta = 2;
}
